docs/continuous-delivery/recipes/terraform/share-common-variables-naming-conventions.md

```diff
@@ -0,0 +1,304 @@
+
+One of the limitation of the variables declaration is that it's not possible to compose variables, [locals](https://www.terraform.io/language/values/locals) or Terraform [built-in functions](https://www.terraform.io/language/functions) are used for that.
+
+### Common Terraform module
+
+One way to bypass this limitations is to introduce a "common" module, that will not deploy any resources, but just compute / calculate and output the resource names and shared variables, and be used by all other modules, as a dependency.
+
+```console
+modules
+├── common
+│   ├── output.tf
+│   └── variables.tf
+├── kubernetes
+│   ├── main.tf
+│   ├── provider.tf
+│   └── variables.tf
+├── network
+│   ├── main.tf
+│   ├── provider.tf
+│   └── variables.tf
+```
+
+*variables.tf:*
+
+```hcl
+variable "environment_name" {
+  type = string
+  description = "The name of the environment."
+}
+
+variable "location" {
+  type = string
+  description = "The Azure region where the resources will be created. Default is westeurope."
+  default = "westeurope"
+}
+```
+
+*output.tf:*
+
+```hcl
+# Shared variables
+output "location" {
+  value = var.location
+}
+
+output "subscription" {
+  value = var.subscription
+}
+
+# Virtual Network Naming
+
+output "vnet_rg_name" {
+  value = "rg-network-${var.environment_name}"
+}
+
+output "vnet_name" {
+  value = "vnet-${var.environment_name}"
+}
+
+# AKS Naming
+
+output "aks_rg_name" {
+  value = "rg-aks-${var.environment_name}"
+}
+
+output "aks_name" {
+  value = "aks-${var.environment_name}"
+}
+```
+
+Now, if you execute the Terraform apply for the common module, you get all the shared/common variables in outputs:
+
+```console
+$ terraform plan -var environment_name="dev" -var subscription="$(az account show --query id -o tsv)"
+
+Changes to Outputs:
+  + aks_name     = "aks-dev"
+  + aks_rg_name  = "rg-aks-dev"
+  + location     = "westeurope"
+  + subscription = "01010101-1010-0101-1010-010101010101"
+  + vnet_name    = "vnet-dev"
+  + vnet_rg_name = "rg-network-dev"
+
+You can apply this plan to save these new output values to the Terraform state, without changing any real infrastructure.
+```
+
+### Use the common Terraform module
+
+Using the common Terraform module in any other module is super easy. For example, this is what you can do in the Azure Kubernetes module `main.tf` file:
+
+```hcl
+module "common" {
+  source           = "../common"
+  environment_name = var.environment_name
+  subscription     = var.subscription
+}
+
+data "azurerm_subnet" "aks_subnet" {
```

Comment:
In the past when there is a single resource of that type, I have seen the usage of the name `this` for the resource name. Do you think this is something that is a convention we should also add to this doc? I am not sure if there is guidance for this from the solution area

Example:
```hcl
data "azurerm_subnet" "this" {
```

Reply:
@omusavi this is a practice that I've seen a very few times in the past, but I find it confusing TBH. In most projects I've seen, data sources and resources that are created inside a module are considered as instance of objects in the module instance itself and named using very explicit nouns like it would have been done in C# or Python, for example.

@hattan does your team have some best practices recommendations for such a case?

Reply:
BTW, according to [this page](https://github.com/microsoft/code-with-engineering-playbook/blob/main/docs/code-reviews/recipes/terraform.md) of the playbook, we shoud follow this [naming convention](https://github.com/jonbrouse/terraform-style-guide/blob/master/README.md#resource-naming) at CSE. And it recommends to use the same name as the type, minus the provider. So in this case, I should have use `subnet`. And `subnet_aks` if multiple subnets were referenced in the module.

Reply:
@omusavi can you share the advantage of this? If it has more advantage than the current naming convention we should consider it

Reply:
no major advantage, but this is the naming convention i have always seen. Thanks for the link to the playbook page with the CSE opinion on convention. I think `this` is nice because it's always consistent, and since you always have to include the resource type anyways, it seems repetitive to me. Thanks for all the links though!